File: app/services/shipstation/order_status.rb
# frozen_string_literal: true

module Shipstation
  # Looks up whether a ShipStation order has already shipped / been labeled, used
  # to decide if a Fluid-side edit may be re-pushed. Fail-safe: on any error or
  # ambiguity it reports `shipped? == true` so we DON'T overwrite an order we
  # can't confirm is still open.
  class OrderStatus < BaseService
    attr_reader :api_key, :api_secret

    def initialize(company_id)
      setting = IntegrationSetting.find_by(company_id: company_id)
      @api_key = setting&.settings&.dig("api_key")
      @api_secret = setting&.settings&.dig("api_secret")
    end

    def shipped?(shipstation_order_id)
      return false if shipstation_order_id.blank? || api_key.blank? || api_secret.blank?

      response = HTTParty.get("#{SHIPSTATION_API_BASE}/orders/#{shipstation_order_id}", headers: headers)
      return true unless response.code == 200 # unsure -> treat as labeled, skip the update

      JSON.parse(response.body)["orderStatus"] == "shipped"
    rescue StandardError => e
      Rails.logger.error("[Shipstation::OrderStatus] #{e.class}: #{e.message}")
      true
    end
  end
end

File: src/lib/shipstation/order-status.ts
/**
 * Port of app/services/shipstation/order_status.rb.
 *
 * Answers whether a ShipStation order has already shipped / been labeled, which
 * is what decides if a Fluid-side edit may be re-pushed. Fail-safe by design: on
 * any error or ambiguity it reports `true`, so we DON'T overwrite an order we
 * cannot confirm is still open.
 */

import { credentialsFor, hasV1Credentials, shipstationGet } from "./client";

export async function isShipstationOrderShipped(
  companyId: bigint,
  shipstationOrderId: string | null,
): Promise<boolean> {
  if (!shipstationOrderId) return false;

  const credentials = await credentialsFor(companyId);
  if (!hasV1Credentials(credentials)) return false;

  try {
    const response = await shipstationGet(
      `/orders/${shipstationOrderId}`,
      credentials,
    );
    // Unsure -> treat as labeled, skip the update.
    if (response.status !== 200) return true;

    const body = (await response.json()) as { orderStatus?: string };
    return body.orderStatus === "shipped";
  } catch (error) {
    console.error(
      "[Shipstation::OrderStatus]",
      error instanceof Error ? error.message : error,
    );
    return true;
  }
}
